File: rtls-react-native/android/src/main/java/com/rtls/reactnative/RTLSyncModule.kt
package com.rtls.reactnative

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import android.util.Log
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.rtls.kmp.BatchingPolicy
import com.rtls.kmp.LocationPoint
import com.rtls.kmp.LocationRequestParams
import com.rtls.kmp.LocationSyncClient
import com.rtls.kmp.LocationSyncClientEvent
import com.rtls.kmp.RTLSKmp
import com.rtls.kmp.SyncEngineEvent
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

private const val EVENT_RECORDED = "rtls_recorded"
private const val EVENT_SYNC_EVENT = "rtls_syncEvent"
private const val EVENT_ERROR = "rtls_error"
private const val EVENT_AUTHORIZATION_CHANGED = "rtls_authorizationChanged"
private const val EVENT_TRACKING_STARTED = "rtls_trackingStarted"
private const val EVENT_TRACKING_STOPPED = "rtls_trackingStopped"

class RTLSyncModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String = "RTLSyncModule"

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private var client: LocationSyncClient? = null
    private var lastUserId: String? = null
    private var lastDeviceId: String? = null
    private var eventsJob: kotlinx.coroutines.Job? = null
    private var locationRequestParams: LocationRequestParams = LocationRequestParams()

    private fun emit(name: String, params: WritableMap?) {
        reactApplicationContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            ?.emit(name, params ?: Arguments.createMap())
    }

    private fun pointToMap(point: LocationPoint): WritableMap {
        val map = Arguments.createMap()
        map.putString("id", point.id)
        map.putString("userId", point.userId)
        map.putString("deviceId", point.deviceId)
        map.putDouble("recordedAt", point.recordedAtMs.toDouble())
        map.putDouble("lat", point.lat)
        map.putDouble("lng", point.lng)
        point.horizontalAccuracy?.let { map.putDouble("horizontalAccuracy", it) }
        point.verticalAccuracy?.let { map.putDouble("verticalAccuracy", it) }
        point.altitude?.let { map.putDouble("altitude", it) }
        point.speed?.let { map.putDouble("speed", it) }
        point.course?.let { map.putDouble("course", it) }
        return map
    }

    private fun startEventCollection(c: LocationSyncClient) {
        eventsJob?.cancel()
        eventsJob = scope.launch {
            c.events.collectLatest { event ->
                when (event) {
                    is LocationSyncClientEvent.Recorded -> {
                        emit(EVENT_RECORDED, pointToMap(event.point))
                    }
                    is LocationSyncClientEvent.SyncEvent -> {
                        val payload = Arguments.createMap()
                        when (val e = event.event) {
                            is SyncEngineEvent.UploadSuccess -> {
                                payload.putString("type", "uploadSucceeded")
                                payload.putInt("accepted", e.accepted)
                                payload.putInt("rejected", e.rejected)
                            }
                            is SyncEngineEvent.UploadFailed -> {
                                payload.putString("type", "uploadFailed")
                                payload.putString("message", e.message)
                            }
                            else -> payload.putString("type", "unknown")
                        }
                        emit(EVENT_SYNC_EVENT, payload)
                    }
                    is LocationSyncClientEvent.Error -> {
                        val payload = Arguments.createMap()
                        payload.putString("message", event.message)
                        emit(EVENT_ERROR, payload)
                    }
                    LocationSyncClientEvent.TrackingStarted -> emit(EVENT_TRACKING_STARTED, Arguments.createMap())
                    LocationSyncClientEvent.TrackingStopped -> emit(EVENT_TRACKING_STOPPED, Arguments.createMap())
                }
            }
        }
    }

    @ReactMethod
    fun configure(config: com.facebook.react.bridge.ReadableMap, promise: Promise) {
        try {
            val baseURL = config.getString("baseURL") ?: run {
                promise.reject("RTLSync", "configure requires baseURL", null)
                return
            }
            val userId = config.getString("userId") ?: run {
                promise.reject("RTLSync", "configure requires userId", null)
                return
            }
            val deviceId = config.getString("deviceId") ?: run {
                promise.reject("RTLSync", "configure requires deviceId", null)
                return
            }
            val accessToken = config.getString("accessToken") ?: run {
                promise.reject("RTLSync", "configure requires accessToken", null)
                return
            }

            val batchMaxSize = if (config.hasKey("batchMaxSize")) config.getInt("batchMaxSize") else null
            val flushIntervalSec = if (config.hasKey("flushIntervalSeconds")) config.getDouble("flushIntervalSeconds") else null
            val maxBatchAgeSec = if (config.hasKey("maxBatchAgeSeconds")) config.getDouble("maxBatchAgeSeconds") else null
            val batchingPolicy = BatchingPolicy(
                maxBatchSize = batchMaxSize?.coerceAtLeast(1) ?: 50,
                flushIntervalSeconds = (flushIntervalSec?.toLong()?.coerceAtLeast(1)) ?: 10L,
                maxBatchAgeSeconds = (maxBatchAgeSec?.toLong()?.coerceAtLeast(0)) ?: 60L
            )

            val intervalSec = if (config.hasKey("locationIntervalSeconds")) config.getDouble("locationIntervalSeconds") else null
            val distanceM = if (config.hasKey("locationDistanceMeters")) config.getDouble("locationDistanceMeters") else null
            val significantOnly = if (config.hasKey("useSignificantLocationOnly")) config.getBoolean("useSignificantLocationOnly") else false
            locationRequestParams = when {
                significantOnly -> LocationRequestParams(
                    intervalMillis = 60_000L,
                    minUpdateIntervalMillis = 60_000L,
                    minUpdateDistanceMeters = 500f
                )
                intervalSec != null && intervalSec > 0 -> {
                    val ms = (intervalSec * 1000).toLong().coerceAtLeast(5_000L)
                    LocationRequestParams(intervalMillis = ms, minUpdateIntervalMillis = ms, minUpdateDistanceMeters = 0f)
                }
                distanceM != null && distanceM > 0 -> LocationRequestParams(
                    intervalMillis = 10_000L,
                    minUpdateIntervalMillis = 5_000L,
                    minUpdateDistanceMeters = distanceM.toFloat().coerceAtLeast(1f)
                )
                else -> LocationRequestParams()
            }

            val ctx = reactApplicationContext
            client?.stopTracking()
            lastUserId = userId
            lastDeviceId = deviceId
            val c = RTLSKmp.createLocationSyncClient(
                context = ctx,
                baseUrl = baseURL,
                userId = userId,
                deviceId = deviceId,
                accessToken = accessToken,
                scope = scope,
                batchingPolicy = batchingPolicy
            )
            client = c
            startEventCollection(c)
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("RTLSyncModule", "configure failed", e)
            promise.reject("RTLSync", e.message, e)
        }
    }

    @ReactMethod
    fun startTracking(promise: Promise) {
        val ctx = reactApplicationContext
        val c = client
        if (c == null) {
            promise.reject("RTLSync", "Call configure first", null)
            return
        }
        val userId = lastUserId ?: run {
            promise.reject("RTLSync", "Configure first", null)
            return
        }
        val deviceId = lastDeviceId ?: run {
            promise.reject("RTLSync", "Configure first", null)
            return
        }
        if (!hasLocationPermission(ctx)) {
            promise.reject("LOCATION_PERMISSION_DENIED", "Location permission not granted. Call requestAlwaysAuthorization() first.", null)
            return
        }
        try {
            val flow = RTLSKmp.createLocationFlow(ctx, userId, deviceId, locationRequestParams)
            c.startCollectingLocation(flow)
            promise.resolve(null)
        } catch (e: Exception) {
            Log.e("RTLSyncModule", "startTracking failed", e)
            promise.reject("RTLSync", e.message, e)
        }
    }

    @ReactMethod
    fun stopTracking(promise: Promise) {
        try {
            client?.stopTracking()
            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("RTLSync", e.message, e)
        }
    }

    @ReactMethod
    fun requestAlwaysAuthorization(promise: Promise) {
        val activity = currentActivity
        if (activity == null) {
            promise.reject("NO_ACTIVITY", "Activity not available to request permission", null)
            return
        }
        val permissions = mutableListOf(
            Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.ACCESS_COARSE_LOCATION
        )
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            permissions.add(Manifest.permission.ACCESS_BACKGROUND_LOCATION)
        }
        ActivityCompat.requestPermissions(activity, permissions.toTypedArray(), REQUEST_CODE_LOCATION)
        promise.resolve(null)
    }

    @ReactMethod
    fun getStats(promise: Promise) {
        val c = client
        if (c == null) {
            promise.reject("RTLSync", "Call configure first", null)
            return
        }
        scope.launch {
            try {
                val stats = c.stats()
                withContext(Dispatchers.Main) {
                    val map = Arguments.createMap()
                    map.putInt("pendingCount", stats.pendingCount)
                    if (stats.oldestPendingRecordedAtMs != null) {
                        map.putDouble("oldestPendingRecordedAt", stats.oldestPendingRecordedAtMs.toDouble())
                    } else {
                        map.putNull("oldestPendingRecordedAt")
                    }
                    promise.resolve(map)
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    promise.reject("RTLSync", e.message, e)
                }
            }
        }
    }

    @ReactMethod
    fun flushNow(promise: Promise) {
        val c = client
        if (c == null) {
            promise.reject("RTLSync", "Call configure first", null)
            return
        }
        scope.launch {
            try {
                c.flushNow()
                withContext(Dispatchers.Main) { promise.resolve(null) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    promise.reject("RTLSync", e.message, e)
                }
            }
        }
    }

    override fun invalidate() {
        eventsJob?.cancel()
        client?.stopTracking()
        client = null
        scope.cancel()
        super.invalidate()
    }

    private fun hasLocationPermission(ctx: android.content.Context): Boolean {
        return ContextCompat.checkSelfPermission(ctx, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED
    }

    companion object {
        private const val REQUEST_CODE_LOCATION = 0x2a03
    }
}

File: rtls-react-native/ios/RTLSyncModule.swift
// React Native native module bridging to RTLSyncKit (iOS).
// The host app must add the Swift package (this repo's Package.swift) in Xcode.

import Foundation
import React
import RTLSyncKit

private let eventRecorded = "rtls_recorded"
private let eventSyncEvent = "rtls_syncEvent"
private let eventError = "rtls_error"
private let eventAuthorizationChanged = "rtls_authorizationChanged"
private let eventTrackingStarted = "rtls_trackingStarted"
private let eventTrackingStopped = "rtls_trackingStopped"

private func locationPointToDict(_ point: LocationPoint) -> [String: Any] {
    var dict: [String: Any] = [
        "id": point.id.uuidString,
        "userId": point.userId,
        "deviceId": point.deviceId,
        "recordedAt": Int(point.recordedAt.timeIntervalSince1970 * 1000),
        "lat": point.coordinate.latitude,
        "lng": point.coordinate.longitude,
    ]
    if let v = point.horizontalAccuracy { dict["horizontalAccuracy"] = v }
    if let v = point.verticalAccuracy { dict["verticalAccuracy"] = v }
    if let v = point.altitude { dict["altitude"] = v }
    if let v = point.speed { dict["speed"] = v }
    if let v = point.course { dict["course"] = v }
    return dict
}

private func authStatusString(_ auth: LocationAuthorization) -> String {
    switch auth {
    case .notDetermined: return "notDetermined"
    case .restricted: return "restricted"
    case .denied: return "denied"
    case .authorizedWhenInUse: return "authorizedWhenInUse"
    case .authorizedAlways: return "authorizedAlways"
    }
}

@objc(RTLSyncModule)
class RTLSyncModule: RCTEventEmitter {

    private var client: LocationSyncClient?
    private var eventTask: Task<Void, Never>?

    override init() {
        super.init()
    }

    override static func requiresMainQueueSetup() -> Bool {
        return false
    }

    override func supportedEvents() -> [String]! {
        return [
            eventRecorded,
            eventSyncEvent,
            eventError,
            eventAuthorizationChanged,
            eventTrackingStarted,
            eventTrackingStopped,
        ]
    }

    private func startEventLoop(client: LocationSyncClient) {
        eventTask?.cancel()
        eventTask = Task { [weak self] in
            guard let self = self else { return }
            for await event in client.events {
                if Task.isCancelled { break }
                await self.emit(event)
            }
        }
    }

    private func emit(_ event: LocationSyncClientEvent) async {
        let (name, body): (String, [String: Any]?) = switch event {
        case .recorded(let point):
            (eventRecorded, locationPointToDict(point))
        case .syncEvent(let e):
            let payload: [String: Any] = switch e {
            case .didUpload(let accepted, let rejected):
                ["type": "uploadSucceeded", "accepted": accepted, "rejected": rejected]
            case .uploadFailed(let msg):
                ["type": "uploadFailed", "message": msg]
            }
            (eventSyncEvent, payload)
        case .error(let msg):
            (eventError, ["message": msg])
        case .authorizationChanged(let auth):
            (eventAuthorizationChanged, ["status": authStatusString(auth)])
        case .trackingStarted:
            (eventTrackingStarted, [:])
        case .trackingStopped:
            (eventTrackingStopped, [:])
        }
        await MainActor.run {
            self.sendEvent(withName: name, body: body ?? [:])
        }
    }

    @objc func configure(_ config: NSDictionary, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let baseURLString = config["baseURL"] as? String,
              let baseURL = URL(string: baseURLString),
              let userId = config["userId"] as? String,
              let deviceId = config["deviceId"] as? String,
              let accessToken = config["accessToken"] as? String else {
            reject("RTLSync", "configure requires baseURL, userId, deviceId, accessToken", nil)
            return
        }
        let token = accessToken
        let provider = AuthTokenProvider { token }
        let databaseURL: URL
        do {
            databaseURL = try LocationSyncClientConfiguration.defaultDatabaseURL(directoryName: "RTLSyncKitRN")
        } catch {
            reject("RTLSync", "Failed to get database URL: \(error)", error as NSError)
            return
        }
        let configuration = LocationSyncClientConfiguration(
            baseURL: baseURL,
            authTokenProvider: provider,
            userId: userId,
            deviceId: deviceId,
            databaseURL: databaseURL
        )
        Task { [weak self] in
            guard let self = self else { return }
            do {
                let c = try await LocationSyncClient(configuration: configuration)
                self.client = c
                self.startEventLoop(client: c)
                await MainActor.run { resolve(NSNull()) }
            } catch {
                await MainActor.run { reject("RTLSync", error.localizedDescription, error as NSError) }
            }
        }
    }

    @objc func startTracking(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let client = client else {
            reject("RTLSync", "Call configure first", nil)
            return
        }
        Task {
            await client.startTracking()
            await MainActor.run { resolve(NSNull()) }
        }
    }

    @objc func stopTracking(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let client = client else {
            reject("RTLSync", "Call configure first", nil)
            return
        }
        Task {
            await client.stopTracking()
            await MainActor.run { resolve(NSNull()) }
        }
    }

    @objc func requestAlwaysAuthorization(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let client = client else {
            reject("RTLSync", "Call configure first", nil)
            return
        }
        Task {
            await client.requestAlwaysAuthorization()
            await MainActor.run { resolve(NSNull()) }
        }
    }

    @objc func getStats(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let client = client else {
            reject("RTLSync", "Call configure first", nil)
            return
        }
        Task {
            let stats = await client.stats()
            let oldestMs: NSNumber? = stats.oldestPendingRecordedAt.map { NSNumber(value: Int($0.timeIntervalSince1970 * 1000)) }
            await MainActor.run {
                resolve([
                    "pendingCount": stats.pendingCount,
                    "oldestPendingRecordedAt": oldestMs ?? NSNull(),
                ])
            }
        }
    }

    @objc func flushNow(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let client = client else {
            reject("RTLSync", "Call configure first", nil)
            return
        }
        Task {
            await client.flushNow()
            await MainActor.run { resolve(NSNull()) }
        }
    }
}
